packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -2,20 +2,158 @@
+  /// constructors for [ErrorSummary], [ErrorDetails], and [ErrorHint] to use
+  /// this constructor.
+  ///
+  /// {@tool sample}
+  /// ```dart
+  /// _ErrorDiagnostic('Element $element must be $color')
+  /// ```
+  /// Desugars to:
+  /// ```dart
+  /// _ErrorDiagnostic._fromParts(<Object>['Element ', element, ' must be ', color]);
+  /// ```
+  ///
+  /// Slightly more complex case:
+  /// ```dart
+  /// _ErrorDiagnostic('Element ${element.runtimeType} must be $color');
+  /// ```
+  /// Desugars to:
+  ///```dart
+  /// _ErrorDiagnostic._fromParts(<Object>[
+  ///   'Element ',
+  ///   DiagnosticsProperty(null, element, description: element.runtimeType?.toString()),
+  ///   ' must be ',
+  ///   color,
+  /// ]);
+  /// ```
+  /// {@end-tool}
+  _ErrorDiagnostic._fromParts(
+    List<Object> messageParts, {
+    DiagnosticsTreeStyle style = DiagnosticsTreeStyle.flat,
+    DiagnosticLevel level = DiagnosticLevel.info,
+  }) : assert(messageParts != null),
+       super(
+         null,
+         messageParts,
+         showName: false,
+         showSeparator: false,
+         defaultValue: null,
+         style: style,
+         level: level,
+       );
+
+  @override
+  String valueToString({ TextTreeConfiguration parentConfiguration }) {
+    return value.join('');
+  }
+}
+
+/// An explanation of the problem and its cause, any information that may help
+/// track down the problem, background information, etc.
+///
+/// Use [ErrorDescription] for any part of an error message where neither
+/// [ErrorSummary] or [ErrorHint] is appropriate.
+///
+/// See also:
+///
+/// * [ErrorSummary], which provides a short (one line) description of the
+///   problem that was detected.
+/// * [ErrorHint], which provides specific, non-obvious advice that may be
+///   applicable.
+class ErrorDescription extends _ErrorDiagnostic {
+  /// This constructor provides a reliable hook for a kernel transformer to find
+  /// error messages that need to be rewritten to include object references for
+  /// interactive display of errors.
+  ///
+  /// The message will display with the same text regardless of whether the
+  /// kernel transformer is used. The kernel transformer is required so that
+  /// debugging tools can provide interactive displays of objects described by
+  /// the error.
+  ErrorDescription(String message) : super._fromParts(<Object>[message], level: DiagnosticLevel.info);
+
+  /// Calls to the default constructor may be rewritten to use this constructor
+  /// in debug mode using a kernel transformer.
+  ErrorDescription._fromParts(List<Object> messageParts) : super._fromParts(messageParts, level: DiagnosticLevel.info);
+}
+
+/// A short (one line) description of the problem that was detected.
+///
+/// Error summaries from the same source location should have little variance,
+/// so that they can be recognized as related. For example, they shouldn't
+/// include hash codes.
```

Comment:
should we say somewhere that the first diagnostic to a FlutterError should be an ErrorSummary? Should we say there should only be one ErrorSummary per FlutterError? Should we asserts those things?

Reply:
I've documented those invariants and added lints  with asserts enforcing them. The asserts all throw FlutterError objects that include the underlying malformed FlutterError along with a description of how the error is malformed so framework authors know what to fix and framework users can still see the slightly malformed but still useful error.